### PR TITLE
Performance improvements for depthwise convolutions in FP16

### PR DESCRIPTION
**Pull request by ptrblck**
This PR activates faster depthwise convolution kernels for Volta and Turing GPUs using cudnn >= 7600.
The script to benchmark the current PyTorch master branch and this PR branch can be found [here](https://gist.github.com/ptrblck/4590cf20721d8f43296c9903abd4a774).
(50 warmup iterations, 1000 iterations for timing)

I've used #3265 to create a similar benchmark and added a few additional setups.
Since the results are quite long, I've uploaded them in a spreadsheet [here](https://docs.google.com/spreadsheets/d/13ByXcqg7LQUr3DVG3XpLwnJ-CXg3GUZJ3puyTMw9n2I/edit?usp=sharing).
Times are given in ms per iteration. 
We've benchmarked this PR on a DGX1 using V100 GPUs.

The current workload check in `check_cudnn_depthwise_workload` is quite long and can be moved to another file, if wanted.

CC @ngimel (Thanks for the support while benchmarking it ;) )

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -187,6 +194,146 @@ auto ConvParams::is_depthwise(
          weight.size(0) % input.size(1) == 0; // output channels must be a multiple of input channels
 }
 
+// Check workload to activate fast depthwise FP16 cudnn conv kernels
+bool check_cudnn_depthwise_workload(const at::Tensor& input, const at::Tensor& weight, int stride) {
+  int w = input.size(3);  // same as h
+  int ch = input.size(1);
+  int bs = input.size(0);
+  int k = weight.size(2); // kernel size
```
Comment (ngimel) [inline aten/src/ATen/native/Convolution.cpp]: you never use `k` in this function

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/Convolution.cpp (hunk 1 above)
Comment (ptrblck) [reply aten/src/ATen/native/Convolution.cpp]: You are right! That's some dead code and I'll remove it.

**Comment by ngimel** [comment]
@ptrblck windows build failure looks real, it does not like "and" apparently.

**Comment by ptrblck** [comment]
@ngimel Thanks for the information!
I wasn't aware that alternative tokens might cause trouble in Windows.
I've updated it to `&&`.

File: aten/src/ATen/native/Convolution.cpp [hunk 2]
```diff
@@ -187,6 +194,145 @@ auto ConvParams::is_depthwise(
...
+    }
+  }
+  return false;
+}
+
+// Use cudnn for FP16 depthwise convolutions
+auto ConvParams::use_cudnn_depthwise(
+        const at::Tensor& input, const at::Tensor& weight) const -> bool {
+  #if AT_CUDNN_ENABLED()
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
```
Comment (ngimel) [inline aten/src/ATen/native/Convolution.cpp]: you should not be calling getCurrentDeviceProperties() directly here, instead add CUDAHooks::supportsDepthwiseConvolutionWithCuDNN to cuda/detail/CUDAHooks.cpp, like it's currently done for CUDAHooks::supportsDilatedConvolutionWithCuDNN(). That would also allow you to not use AT_CUDNN_ENABLED macro.

**Comment by yaysummeriscoming** [comment]
Hi I’m very excited this is in :) . I’m able to reproduce the individual depthwise convolution tests as presented in the spreadsheet - very impressive gains of up to 400%.  

I decided to test this with MobileNet V2, however unfortunately I’m only seeing speedups of ~10%.  My understanding is that depthwise convolution is the slowest link when training such lightweight networks, so this seems quite low to me?

I’ve modified the test script to use MobileNet V2 here:
https://gist.github.com/yaysummeriscoming/88ae59bc5b7ba8581ea396d8ce87d28f

I had a go at profiling with the autograd profiler, but that doesn’t delineate between point wise & depthwise convolution.

Any pointers?  Could it be that cuDNN isn’t optimised for pointwise convolutions with large input/output channel ratios?

(I’m assuming this is the right place to ask this, please correct me if not)

**Comment by jph00** [comment]
@ptrblck thanks for this great PR and helpful benchmarking. I've created a couple of summary tables of the benchmarks that might be helpful. Here's speedup by kernel size and stride, by height/width:

image

And here's the details of h/w by num channels, for just the stride one and kernel size 3 rows:

image

Have you tried benchmarking 5x5 convs? They are used a lot in efficientnet so would be great if they're fast...

cc @ngimel

**Comment by jph00** [comment]
I just tried 5x5 convs and it appears they are *not* optimized for tensor cores - they ran at about the same speed for fp16 vs fp32.

**Comment by bearpelican** [comment]
Looks like cudnn is only enabled for 1x1 and 3x3 - https://github.com/pytorch/pytorch/pull/22302/files#diff-57ac615408468d3c7a461e505581bea3R316. 

Would enabling 5x5 change anything?

**Comment by ngimel** [comment]
IIRC, cudnn only had fast implementations for 1x1 and 3x3, so just enabling it for 5x5 is unlikely to dramatically speed things up.

**Comment by jph00** [comment]
Thanks @ngimel . Is there any plan to add a 5x5 implementation? If not, could I twist your arm to create such a plan... ;)

**Comment by ngimel** [comment]
Not mine, I'm not with nvidia anymore :-)

**Comment by jph00** [comment]
Oh yes so I see! Welcome to Facebook then :)

**Comment by andravin** [comment]
Hi @ptrblck and @ngimel , just to clarify a point that is causing some confusion: although this patch uses cuDNN kernels for depthwise convolution on Volta, do those kernels actually use tensor cores? 

Depthwise convolution is basically planar convolution nested inside of diagonal matrix multiplication. I do not see how that could be made faster with `8m x 8n x 4k` matrix multiplication fragments. It seems like one of the input matrices would be diagonal, and (at most) 1/8th of the matrix elements would be nonzero, reducing the effective arithmetic throughput of the tensor cores to the level of fp32 core arithmetic throughput.

**Comment by ngimel** [comment]
@andravin you are right, those kernels don't use tensor cores.

**Comment by andravin** [comment]
@ngimel do they use `hfma2`?  fp16 accumulation might be OK for `3x3` depth-wise.

Otherwise I am stumped why these kernels are Volta only. Also, P100 had `hfma2`.

**Pull request closed without merging**